src/main/java/org/sonatype/repository/conan/internal/proxy/ConanProxyFacet.java

```diff
@@ -93,6 +111,21 @@ protected void doValidate(final Configuration configuration) throws Exception {
     super.doValidate(configuration);
   }
 
+  @Override
+  protected void doConfigure(final Configuration configuration) throws Exception {
+    config = facet(ConfigurationFacet.class).readSection(configuration, CONFIG_KEY, Config.class);
+
+    urlMatcher.accept(
```

Comment:
@fjmilens3 You are correct and this is how I originally started to write the code but I thought that it made it a little difficult to follow with if/else based on config.
What if a third combination of path token is introduced by another server? I thought this approach was a cleaner separation of concerns.
If it is OK I will wait for others to comment before reverting the changes

Reply:
> You are correct and this is how I originally started to write the code but I thought that it made it a little difficult to follow with if/else based on config.

To me it's harder to follow an abstract matcher class, two concrete implementations that only differ on the ordering of three path parts, and a custom JSON deserializer that's used to instantiate a matcher as part of a config object that gets passed into the format recipe. If you push the logic down into the path parsing and path building then I think you get a lot of the same benefits without so much of the code bloat.

> What if a third combination of path token is introduced by another server? I thought this approach was a cleaner separation of concerns.

I'm not that worried if they have additional combinations of path tokens because there are a variety of ways you could deal with that. Probably the easiest would be:

```
switch (gavOrdering) {
case GAV:
  return new Coord(part1(tokens), part2(tokens), part3(tokens));
case AGV:
  return new Coord(part2(tokens), part1(tokens), part3(tokens));
case VGA:
  return new Coord(part3(tokens), part1(tokens), part2(tokens));
case GVA:
  return new Coord(part1(tokens), part3(tokens), part2(tokens));
case AVG:
  return new Coord(part2(tokens), part3(tokens), part1(tokens));
case VAG:
  return new Coord(part3(tokens), part2(tokens), part1(tokens));
}
```

If the number just stays the same there's only the six possible permutations for the ordering, and I'd argue a switch statement with six cases is still more maintainable than six different matcher implementations that vary based only on the parts of those routes.

Now, if you run into one that has a fundamentally different _route_ rather than just the _ordering_ then you have a stronger claim, but even then, I might just worry about that when and if that ever happens, or add additional routes to cover that particular exception to the rule. There's no telling what other weird thing will happen in the future, unfortunately, and most of the stuff we do with formats is very much a living document.

> If it is OK I will wait for others to comment before reverting the changes

Of course that's okay. I'd like to see what everyone else has to say as well, particularly since you and I talked about this together but never really opened it to the broader group. Even if they tend to agree more with me I think it's good that we see that there is, in fact, a way to work around and customize the matchers in this way in case we ever find out that we need to do that. I'd never have thought to do it this way.

Reply:
I have raised this as a separate ticket so that it can be addressed at a later time https://github.com/sonatype-nexus-community/nexus-repository-conan/issues/16

Reply:
I'm late to the party here but I just wanted to add that @fjmilens3 is a lot easier to understand/maintain (I like the introduction of GAV/AGV etc). Only thing I'd add is that GAV might only be understood by Maven/Sonatype developers so I'd either add a comment that translates it to Conan terminology or preferably use Conan terminology in place of GAV. 

Also.... @fjmilens3 your last case statement... 😆